### PR TITLE
Code samples usage

### PR DESCRIPTION
**Pull request by adriendupuis**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | N/A
| Versions      | master, 4.6
| Edition       | N/A

Tool to search and preview code samples usage.
It outputs a list of Markdown code blocks using `include_file`, each followed by a preview of its rendering.
For examples:
- `php tools/code_samples/code_samples_usage.php code_samples/api/product_catalog/src/Command/AttributeCommand.php code_samples/api/product_catalog/src/Command/ProductCommand.php` lists usages of the two files passed as arguments (argument count is unlimited).
- `php tools/code_samples/code_samples_usage.php` lists every `include_file` usage.

`⫸` indicates a highlighted line (`hl_lines`).

The tool is used by a GitHub action for CI on pull request. It compares the usage of modified `code_samples/` files between the target branch and the current branch. If there is changes in `code_samples/`, this action posts its report as a comment on the PR. It deletes its previous comment if it exists. (It could have edit and update the previous one but an old comment can become hidden by default and stay hidden even if updated.) Because of GitHub limitation, the comment can be hard to read. A colorized version is available as HTML file in a downloadable archive (due to artefact limitation, it can't be seen online or downloaded unzipped).

See the "code_samples/ change report" comment below for an example with two bonus modification.

Bonus:

Two PHP files are modified for demo, they had too long `__construct` lines and new lines are added:
- code_samples/api/product_catalog/src/Command/AttributeCommand.php
- code_samples/api/product_catalog/src/Command/ProductCommand.php

A report comment helps to check side effects on their usage.

The page using them doesn't change:
- https://doc.ibexa.co/en/latest/pim/product_api/
- https://ez-systems-developer-documentation--2438.com.readthedocs.build/en/2438/pim/product_api/

docs/pim/price_api.md has been modified. As the PHP code doesn't change, there is nothing about it in the report comment. Locally, the changes can be checked with `php tools/code_samples/code_samples_usage.php code_samples/api/product_catalog/src/Command/ProductPriceCommand.php` which could be simpler than running mkdocs server.

Remove duplicated blank lines before `// ...`:
- https://doc.ibexa.co/en/latest/pim/price_api/#prices
- https://ez-systems-developer-documentation--2438.com.readthedocs.build/en/2438/pim/price_api/#prices

### Cases checks

- [x] There is no change in code_samples/
- [x] A code_samples/ file have been modified and its inclusions have different outputs than previously
- [x] A code_samples/ file have been modified but it makes no difference in its inclusions
- [x] A file as been renamed in code_samples/ but not in its inclusion
- [x] A file as been renamed in code_samples/ and in its inclusion
- [x] A file as been removed from code_samples/ but it's still included

#### Checklist

- [ ] Text renders correctly
- [ ] Text has been checked with vale
- [ ] Description metadata is up to date
- [ ] Redirects cover removed/moved pages
- [ ] Code samples are working
- [ ] PHP code samples have been fixed with PHP CS fixer
- [ ] Added link to this PR in relevant JIRA ticket or code PR

### CODE REVIEW

**Comment by mnocon** [comment]
<h1>Test</h1>
<table>
<tr style="background-color:powderblue;">
<th>H1</th>
<th>H2</th>
</tr>
<tr style="background-color: blue">
<td>H1</td>
<td>H2</td>
</tr>
</table>

File: tools/code_samples/code_samples_usage.php [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+
+// FRAMEWORK
+
+/**
+ * @param string $targetBranch Branch to compare to.
+ * @param string $sourceBranch Branch to compare from. By default, the current branch HEAD
+ * @return array<int, string> List of file paths from code_samples/ directory.
+ */
+function getModifiedCodeSampleFileList(string $targetBranch = 'origin/master', string $sourceBranch = 'HEAD'): array
```
Comment (mnocon) [inline tools/code_samples/code_samples_usage.php:55]: Is this function used? I don't see the usage

File: tools/code_samples/code_samples_usage.php [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+ * @param string $codeSampleFilePath The path from project root to an included file.
+ *                                  If given, files including this one are returned.
+ *                                  If null, all files including a file are returned.
+ * @return array<int, string> List of file paths from docs/ directory.
+ */
+function getIncludingFileList(string $codeSampleFilePath = null): array
+{
+    $pattern = null === $codeSampleFilePath ? '= include_file' : $codeSampleFilePath;
+    $command = "grep '$pattern' -Rl docs | sort";
+    exec($command, $rawIncludingFileList, $commandResultCode);
```
Comment (mnocon) [inline tools/code_samples/code_samples_usage.php:75]: I'm not sure there is an issue here - but just to be sure that there's no chance of command injection with some crazy filenames - could you use the Process component (https://symfony.com/doc/current/components/process.html)?

I'd also limit the scope of the token to the minimum (https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_token#defining-access-for-the-github_token-permissions) for this job, just for safety

**Review by mnocon — COMMENTED** [review]
Thank you, this is much needed!

I'm just paranoid and I'm wondering if we can harden this a bit, just to be sure - but I really like this.

File: tools/code_samples/code_samples_usage.php (hunk 1 above)
Comment (adriendupuis) [reply tools/code_samples/code_samples_usage.php:55]: You're right. It became unused somewhen in the writing, I had to move to Github PR API. I'll clean this up.

**Comment by adriendupuis** [comment]
Another test in #2485

File: tools/code_samples/code_samples_usage.php (hunk 1 above)
Comment (adriendupuis) [reply tools/code_samples/code_samples_usage.php]: Removed in d09fd94d081b88183f01f8ff29cf14cb7554acb4

File: tools/code_samples/code_samples_usage.php (hunk 2 above)
Comment (adriendupuis) [reply tools/code_samples/code_samples_usage.php:61]: For now I used [`escapeshellarg`](https://www.php.net/manual/function.escapeshellarg.php) around the file path in ebc14f5e2b52b8924d0b9e763ad66478dd00f883.

**Pull request merged**